packages/jaeger-ui/src/model/ddg/transform-ddg-data.tsx

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {
+  TDdgPayload,
+  TDdgService,
+  TDdgOperation,
+  TDdgPath,
+  TDdgServiceMap,
+  TDdgPathElemsByDistance,
+  TDdgTransformedDdgData,
+  PathElem,
+} from './types';
+
+export default function transformDdgData(
+  payload: TDdgPayload,
+  { service: focalService, operation: focalOperation }: { service: string; operation?: string }
+): TDdgTransformedDdgData {
+  const serviceMap: TDdgServiceMap = new Map();
+  const pathElemsByDistance: TDdgPathElemsByDistance = new Map();
+
+  const paths = payload.map(payloadPath => {
+    const path = {} as TDdgPath;
+
+    const members = payloadPath.map(({ operation: operationName, service: serviceName }, i) => {
+      if (!serviceMap.has(serviceName)) {
+        serviceMap.set(serviceName, {
+          name: serviceName,
+          operations: new Map(),
+        });
+      }
+      const service = serviceMap.get(serviceName) as TDdgService;
+
+      if (!service.operations.has(operationName)) {
```

Comment:
A way to avoid extra checks and `as`:

```tsx
let service = serviceMap.get(serviceName);
if (!service) {
    service = {
        name: serviceName,
        operations: new Map(),
    };
    serviceMap.set(serviceName, service);
}
console.log(service.name); // no error - service is known to be non-null
```

The same approach can be used with `operation`.

Reply:
Nice, I will out this approach.